### PR TITLE
Define more clearly how fragments are combined.

### PR DESCRIPTION
**Pull request by toddjonker**
Add a new outermost form, `document`, and new abstract fragment `ivm`. Together, these let us define more precisely how `ion_1_*` works.

*Description of changes:*

This is primarily for discussion; I think the document needs some more work (eg the new prose talks about AST fragments before those are discussed) but I want to get the core ideas out there.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: conformance/README.md [hunk 1]
```diff
@@ -172,66 +172,106 @@ ensuring that all input forms handle scenarios the same way.
...
 ```
-(each (text)
-      (text "$ion_1_0")
-      (binary "E00100EA")
-      (text "$ion_1_1")
-      (binary "E00101EA")
-      (then _form_ ...))
+(document
+  (then (ivm 1 0) _form_ ...)
+  (then (ivm 1 1) _form_ ...))
```
Comment (popematt) [inline conformance/README.md:266]: Still equivalent to this, right?
```
(document (each (ivm 1 0) 
                (ivm 1 1) 
                (then _form_ )))
```

File: conformance/README.md [hunk 2]
```diff
@@ -172,66 +172,106 @@ ensuring that all input forms handle scenarios the same way.
...
-In theory, we can mix and match text, binary, and AST in a single
+In theory, we could mix and match text, binary, and AST fragments in a single
 document, since ultimately they all express the same thing in different ways.
-This means the test-runner cannot simply concatenate fragments, and it must
-either switch parsing modes mid-stream (theoretically sound since fragments must
-contain full top-level values), or transcribe the fragment into one format.
-But there are benefits to this approach:
+However, that forces fairly complicated transcoding into the test framework.
+To keep things reasonable, we constrain test trees such that `text` and `binary`
+fragments cannot coexist on the path to ane expectation.
```
Comment (popematt) [inline conformance/README.md:184]:
```suggestion
fragments cannot coexist on the path to one expectation.
```

File: conformance/README.md [hunk 3]
```diff
@@ -561,12 +601,14 @@ type.
 These rules describe the overall shape of test cases:
 
 ```ebnf
-test ::=  "("  "ion_1_0"  fragment*  continuation  ")"
+test ::=  "("  "document"            continuation  ")"
+       |  "("  "ion_1_0"  fragment*  continuation  ")" 
        |  "("  "ion_1_1"  fragment*  continuation  ")"
        |  "("  "ion_1_x"  fragment*  continuation  ")"
```
Comment (popematt) [inline conformance/README.md:607]: How important is it to retain all these? I'm mostly indifferent right now, but if there were any more, I think it would be too many ways to express the same thing.

File: conformance/README.md (hunk 1 above)
Comment (toddjonker) [reply conformance/README.md:266]: Not quite: no fragment is required by `ion_1_x` but one _is_ required by `then`.  That would cause your proposed expansion to fail.

Why does `then` require a fragment? I want all extensions to be consistent whether they require fragments, and `each` makes no sense without at least one (else it would be effectively an empty-loop).  The need for no-fragment `then` seems limited, so I went for a consistent grammar at the cost of slightly more code for an edge case.

File: conformance/README.md [hunk 4]
```diff
@@ -561,12 +601,14 @@ type.
 These rules describe the overall shape of test cases:
 
 ```ebnf
-test ::=  "("  "ion_1_0"  fragment*  continuation  ")"
+test ::=  "("  "document"            continuation  ")"
```
[the remark this answers is not in the document]
Comment (toddjonker) [reply conformance/README.md:604]: I wanted to (a) keep the grammar simple, and (b) repair a discontinuity: `ion_1_0` etc do not require a fragment, but extensions (`then`/`else`) do require at least one.

I don't expect `document` to be used except for a small number of cases that care about the specific presence, absence, or form of the leading IVM.

[the remark this answers is not in the document]
Comment (toddjonker) [reply conformance/README.md:604]: Put another way; if `document` directly allowed fragments then that would be _more_ behavior to define/implement/test.  Here we delegate that down to existing forms, at the cost of slightly more code for a small number of test cases.

File: conformance/README.md (hunk 3 above)
Comment (toddjonker) [reply conformance/README.md:607]: Since the expansions here are non-trivial (especially `ion_1_x`), I think these are valuable.

I don't see this as many ways to express the same thing, I see this as abstraction over the most common patterns in the test cases, to reduce bulk and potential for error.

File: conformance/README.md [hunk 5]
```diff
@@ -172,66 +172,106 @@ ensuring that all input forms handle scenarios the same way.
 
 ### Combining Formats
 
-TODO: What does combining text + binary mean?
+The previous section implies that in general, a test case can be though of as a
```
Comment (tgregg) [inline conformance/README.md:175]:
```suggestion
The previous section implies that in general, a test case can be thought of as a
```

**Review by tgregg — COMMENTED** [review]
This looks fine to me.

File: conformance/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (toddjonker) [reply conformance/README.md:604]: Hm, actually with `document` present we could turn the `ion_*` clauses into extensions and require at least one fragment.

File: conformance/README.md (hunk 1 above)
Comment (popematt) [reply conformance/README.md:266]: Okay, so it's equivalent to one of these, depending on the remainder of the test case?
```
(document (each (ivm 1 0) 
                (ivm 1 1) 
                (then ... )))

(document (each (ivm 1 0) 
                (ivm 1 1) 
                (each ... )))


(document (each (ivm 1 0) 
                (ivm 1 1) 
                (_expectation_ ... )))
```

Comment (toddjonker) [reply conformance/README.md:266]: I'm a little unclear on what you mean, but I think so? There's always multiple ways to write things.
 
The expansion in the text is something that works regardless of whether the `ion_1_*` body starts with a fragment.  I want to avoid a design that requires inspection of the body.

Comment (toddjonker) [reply conformance/README.md:266]: Here's a context-free expansion of `(ion_1_x <body>)` using each:

```
(document (each (ivm 1 0) 
                (ivm 1 1) 
                (then (toplevel) <body>)))
```

Note the injection of an empty fragment to ensure the `then` clause always has one even if <body> is just an expectation.

Comment (popematt) [reply conformance/README.md:266]: > I want to avoid a design that requires inspection of the body.

Okay, so I think there may be a fundamental difference between how we are approaching this. In your case, this DSL _is_ actually code, and you can do a (presumably) simple transformation. I am not doing any rewriting of the DSL because at the end of the day, this DSL is not a runnable thing in Java—I'm working with `IonReader` or `IonValue` and it needs to be converted or interpreted somehow—so I'm doing the expansion at the same time as I am converting it.

Comment (toddjonker) [reply conformance/README.md:266]: That's not the case; my implementation is a straightforward CS200-level interpreter.  As the tests are being evaluated, it walks the S-expression tree recursively.  There's functions to recognize a "clause" (S-expression starting with a keyword), and a primitive dispatch mechanism using maps from keywords to evaluation functions. There's basically one dispatch map per nonterminal in the grammar, and an evaluation method per nonterminal-option.  There's no preprocessing of the clauses, everything happens on the fly.

The spec doesn't require any implementation to rewrite expressions, its simply defining the semantics of some features like `$ion_1_x` in terms of others in order to be clear and precise.

Side observation, now that I started implementing `denotes`: the more every grammar is a clause, the less code there is in the implementation.

**Review by popematt — COMMENTED** [review]
I'm generally in favor of this change.

I finally think I figured out how to express my reasoning why `(document ...` should accept `fragment*`, so I've written a somewhat lengthy comment about it. However, my intention is not to block this PR until that conversation is resolved. If you want to move ahead without completely resolving that conversation, we can have that conversation in an issue instead.

File: conformance/README.md (hunk 1 above)
Comment (toddjonker) [reply conformance/README.md:266]: Great!  You came to the same conclusion I did regarding what's most sensible.

File: conformance/README.md [hunk 6]
```diff
@@ -172,66 +172,106 @@ ensuring that all input forms handle scenarios the same way.
...
+AST fragments, however, can mix with text or binary fragments.
+This is the case in most situations, since the `ion_1_*` clauses inherently
+abstract over the bytes on an IVM.
+
+Before checking an expectation, the conformance framework constructs a document
+by combining the preceding sequence of fragments:
+  * If any fragment is `text`, any AST fragments are effectively converted to
+text before processing.  Adjacent text fragments MUST be joined with whitespace.
+  * If any fragment is `binary`, any AST fragments are effectively encoded to
+binary before processing.
```
[the remark this answers is not in the document]
Comment (toddjonker) [reply conformance/README.md:194]: As the fragments are being traversed, the Ion encoding is always knowable.  At `document` it's 1.0 by default.  After each fragment it's still known, or at least known to the writer and an implementation could (if not should) surface it via the writing APIs.  I would like to assume that the framework can leverage that.

But I'm also not sure why the test framework needs to care; the "which version" is known to the writer it's transcribing into.

Probably it goes without saying but I think that both `text` and `bytes` should be handled, as much as possible, as direct byte-copy.  `byte`s in particular should not be modified during document aggregation; that would defeat the purpose.

[the remark this answers is not in the document]
Comment (toddjonker) [reply conformance/README.md:194]: Hm, I wonder if the issue is that the writing APIs assume that the Ion-version is determined before creation, and cannot be changed mid-stream.  I would call that an insufficient API, since the Ion format allows version-switching in the midst of a document/stream, so the writers should allow that as well.

File: conformance/README.md [hunk 7]
```diff
@@ -561,14 +601,16 @@ type.
 These rules describe the overall shape of test cases:
 
 ```ebnf
-test ::=  "("  "ion_1_0"  name-string?  fragment*  continuation  ")"
+test ::=  "("  "document" name-string?             continuation  ")"
```
[the remark this answers is not in the document]
Comment (toddjonker) [reply conformance/README.md:604]: I have not intended to express that `document` cannot work with fragments in its grammar rule.  [What I said](https://github.com/amazon-ion/ion-tests/pull/98#discussion_r1628563605) was that doing so leads to more semantics to define here, and more for each implementation to build and test.  The benefit is slightly less syntax for test-cases that use `document`, but I don't think there's any reason to use `document` except for a small handful of cases to test that a no-IVM document behaves the same as if `$ion_1_0` were present. 

For the most part, this entire grammar clause is pedagogical, to enable a crisp and precise definition of the other three root forms.  It also enables the changes to `ion_1_0` such that it does 50% less work than before, and `ion_1_x` does 1/3 less work, because they don't need to cover the no-IVM case through the entire test suite.

I aim to design languages from as few primitives as possible, because that's what must be most carefully specified, and represent the primary parts to build.  The definitions here are such that
* there is a single "bootstrap" form
* that form does one thing (create an empty document)
* all of the semantics of fragment assembly is delegated to (ie reused from) the definition of a continuation.

In other words, this design is very DRY, at the cost of more verbiage for one or two test files total.

File: conformance/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (popematt) [reply conformance/README.md:194]: > At document it's 1.0 by default

I think that this is enough to answer my original question.

> both text and bytes should be handled, as much as possible, as direct byte-copy

Naturally.

>  I would call that an insufficient API, since the Ion format allows version-switching in the midst of a document/stream, so the writers should allow that as well.

If you want to switch versions mid-stream while writing, `ion-java` currently requires you to create a new writer and append the second stream to the first one. We think this is sufficient because most real-world use cases probably don't need to _write_ multiple versions to one stream. Our reader _can_ switch versions mid stream.

I can figure out how to make it work with IVMs that occur in the middle of a fragment.

**Review by popematt — APPROVED** [review]

File: conformance/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (toddjonker) [reply conformance/README.md:604]: After building this change I came to the conclusion that the implementation delta was near trivial to allow more flexibility.  See https://github.com/amazon-ion/ion-tests/pull/108 for changes that I think align with your thinking.
